### PR TITLE
Fix hands on quest 3 and fix eye-gaze fallbacks

### PR DESCRIPTION
**Pull request by AMollis**
# Overview

## Hand Tracking on Quest

Hand pose and selection was not working on Quest 3 when the "Meta Quest Touch Pro Controller Profile" was added a long side the Microsoft MR OpenXR Plugin "Hand Interaction" feature group.   This is because the rotation and position action were bound to the "meta quest controller" even though the controllers weren't active.  Adding checks for an "activeControler" fixes this.

Also, if the `HandPoseDriver` fallbacks to using joint data directly, the tracking state needed to be updated accordingly.  

For the pinch select input reader, the only stable way of determining if a pinch should be driven the raw joint data is by directly querying the HandPoseDriver.  The PinchInputReader now asks the HandPoseDrive if it is using raw joint data for poses.
## Gaze Fallbacks

Also, while fixing this on Quest, I noticed the head gaze fallback code wasn't working.   There were two places with this fallback.  A simple one in the "palm up solver", and the more generic one in TrackedPoseDriverWithFallback.   

The `TrackedPoseDriverWithFallback` wasn't fallbacking to head gaze because of some misplaced NOTs and ORs.  I fixed this and cleaned up the code a bit.   

The "palm up solver" wasn't computing tracking state like general "tracked pose driver", so this could lead to bugs (or inconsistencies).  Also, the "palm up solver" wasn't using the proper transform.

I also fixed the eye gaze tests.  There was a test that was made to test the incorrect behavior, I updated this.  Then there were tests that were commented out because they were failing due to buggy code.  I've reenabled these tests.

## Validated

- HoloLens 2 (with and without calibrated eyes)
- Quest 3 (with and without controllers)

## FIxes

- https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/766

### CODE REVIEW

**Comment by AMollis** [comment]
If anyone is able to test on Quest 3 today/tomorrow, let me know.  Otherwise I will test Friday

File: org.mixedrealitytoolkit.input/Tracking/TrackedPoseDriverExtensions.cs [hunk 1]
```diff
@@ -84,14 +89,27 @@ internal static InputTrackingState GetInputTrackingStateNoCache(this TrackedPose
...
             if (trackingStateAction.HasAnyControls())
             {
-                return (InputTrackingState)trackingStateAction.ReadValue<int>();
+                result = (InputTrackingState)trackingStateAction.ReadValue<int>();
             }
             else
             {
-                return InputTrackingState.None;
+                result = InputTrackingState.None;
             }
```
Comment (keveleigh) [inline org.mixedrealitytoolkit.input/Tracking/TrackedPoseDriverExtensions.cs:100]: nit: this `else` could be removed now that `result` is already initialized to `InputTrackingState.None`

**Review by keveleigh — COMMENTED** [review]
No Quest to help test, unfortunately

**Comment by MaxPalmer-UH** [comment]
I'll try to test this today.

**Comment by MaxPalmer-UH** [comment]
With this branch I see rays when using controllers appear above / below the horizon, as expected. I'd like to double check which options should be on for hand trackling + MRTK + Quest3 though? Strangly, if I enable hand tracking in the OpenXR list (which I assume should be off), I see the issue with rays disappaering when angled below the 'horizon'.

image

Which options do you believe should be enabled for Meta Quest hand tracking on Quest 3 with MRTK3?

When I drop the controllers and it switches to hands, I see no hand rays at all.

**Comment by MaxPalmer-UH** [comment]
With that option on, controllers and tracked hands both show rays that diappear below the horizon.
With than option off, controller show rays all the time, but hands show no rays.
That's with your branch/fork @AMollis 
Also, can I check which subsystems should be enabled under MRTK too please? This is what I'm using:
**
image
**

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs [hunk 2]
```diff
@@ -354,6 +356,39 @@ public void StartWorldLockReattachCheckCoroutine()
...
             new ProfilerMarker("[MRTK] HandConstraintPalmUp.TryGenerateHandPlaneAndActivationPoint");
 
+        /// <summary>
+        /// Get the input tracking state.
+        /// </summary>
+        private InputTrackingState GetInputTrackingState(TrackedPoseDriver driver)
+        {
+            // Note, that the logic in this class is meant to reproduce the same logic as the base. The base
+            // `TrackedPoseDriver` also sets the tracking state in a similar manner. Please see 
+            // `TrackedPoseDriver::ReadTrackingState`. Replicating this logic in a subclass is not ideal, but it is
```
Comment (AMollis) [inline org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs:366]: I should update this comment

File: org.mixedrealitytoolkit.input/Tracking/HandPoseDriver.cs [hunk 3]
```diff
@@ -71,16 +71,24 @@ protected override void PerformUpdate()
...
+                (trackingType == TrackingType.RotationAndPosition || trackingType == TrackingType.PositionOnly) &&
+                (positionInput.action == null || !positionInput.action.HasAnyControls() || positionInput.action.activeControl == null);
+
+            bool missingRotationController =
+                (trackingType == TrackingType.RotationAndPosition || trackingType == TrackingType.RotationOnly) &&
+                (rotationInput.action == null || !rotationInput.action.HasAnyControls() || rotationInput.action.activeControl == null);
 
             // We will also check the tracking state here to account for a bound action but untracked interaction profile.
-            if ((missingPositionController || missingRotationController || IsTrackingNone()) &&
+            if ((missingPositionController || missingRotationController || m_trackingState == InputTrackingState.None) &&
```
Comment (AMollis) [inline org.mixedrealitytoolkit.input/Tracking/HandPoseDriver.cs:88]: should un do this change

**Comment by AMollis** [comment]
> With that option on, both controllers and tracked hands show rays that disappear below an arbitrary horizon. With than option off, controller show rays all the time, but hands show no rays. That's with your branch/fork @AMollis on a Quest3. Also, can I check which subsystems should be enabled under MRTK too please? This is what I'm using: ** image **

Yeah my fix was WAY OFF...  I got my hands on a Quest, and have made a real fix :)

**Comment by MaxPalmer-UH** [comment]
> > With that option on, both controllers and tracked hands show rays that disappear below an arbitrary horizon. With than option off, controller show rays all the time, but hands show no rays. That's with your branch/fork @AMollis on a Quest3. Also, can I check which subsystems should be enabled under MRTK too please? This is what I'm using: ** image **
> 
> Yeah my fix was WAY OFF... I got my hands on a Quest, and have made a real fix :)

Yes, can confirm, now working for me too. Thanks.

**Review by shaynie — APPROVED** [review]

File: org.mixedrealitytoolkit.input/Tracking/TrackedPoseDriverWithFallback.cs [hunk 4]
```diff
@@ -71,41 +69,59 @@ protected override void PerformUpdate()
...
+                hasRotationFallbackAction &&
+                fallbackInputTrackingState.HasFlag(InputTrackingState.Rotation))
             {
-                neededToGetFallbackData = true;
+                fallbackDataUsed |= InputTrackingState.Rotation;
                 rotation = fallbackRotationAction.action.ReadValue<Quaternion>();
             }
 
-            if (neededToGetFallbackData) //because either position, rotation, or both data were obtained from fallback actions
+            // Id either position, rotation, or both data were obtained from fallback actions,
```
Comment (ms-RistoRK) [inline org.mixedrealitytoolkit.input/Tracking/TrackedPoseDriverWithFallback.cs:115]:
```suggestion
            // If either position, rotation, or both data were obtained from fallback actions,
```

**Review by ms-RistoRK — DISMISSED** [review]

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs [hunk 5]
```diff
@@ -294,16 +295,17 @@ private bool IsUserGazeMeetingThresholdRequirements(XRNode hand)
                     }
                 }
                 #pragma warning restore CS0618
-                else if (TrackedPoseDriverLookup != null)
+                else if (TrackedPoseDriverLookup != null &&
+                    TrackedPoseDriverLookup.GazeTrackedPoseDriver != null)
```
Comment (whebertML) [inline org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs:299]: Putting the check for GazeTrackedPoseDriver != null here, causes the logic to skip the fallback to headpose using camera transform.

Perhaps include a null check of the driver inside GetInputTrackingState, where it just returns InputTrackingState.None in that case, allowing the fallback.

**Comment by whebertML** [comment]
@AMollis   Testing with Quest 3.

Seeing an issue of no fallback for HandMenu if the Gaze Controller TrackedPoseDriver isn't defined in TrackedPoseDriverLookup.  That should fallback to using headpose via camera transform.

Also, I don't appear to be getting Select values when switching to the Quest 3 controller in the new rig.  The legacy rig still works fine when using controller, but I don't see select value change when pulling trigger in new rig.  Here is my OpenXR config for reference - was seeing this with either the "Meta Quest Touch Pro Controller Profile" or the "Oculus Touch Controller Profile".

**Comment by AMollis** [comment]
> @AMollis Testing with Quest 3.
> 
> Seeing an issue of no fallback for HandMenu if the Gaze Controller TrackedPoseDriver isn't defined in TrackedPoseDriverLookup. That should fallback to using headpose via camera transform.
> 
> Also, I don't appear to be getting Select values when switching to the Quest 3 controller in the new rig. The legacy rig still works fine when using controller, but I don't see select value change when pulling trigger in new rig. Here is my OpenXR config for reference - was seeing this with either the "Meta Quest Touch Pro Controller Profile" or the "Oculus Touch Controller Profile".
> 
> image

@whebertML  your settings should look like this.  I don't think Query 3 supports the EXT_Hand_Interaction.  So fallbacks using the joint data from Unity's "Hand Tracking Subsystem" feature, Controller data from our "Hand Tracking" feature, Controller models from our "Motion Controller Model"

image

And I find better joint positioning when using Unity's Hand Package

image


Also....the default MRTK3 input actions don't have bindings for those other "Meta" features....so not entirely sure what those will do.

**Comment by whebertML** [comment]
@AMollis 

Interesting, I enabled some of the features like you had, but no difference for me - still no controller selection, though it does track.  Again, the controller totally works with the legacy rig, but for some reason not the new rig, for me.  So you are seeing controller selection working in the new rig when switching from hands to controller?

image

I wonder if it has something to do with using a different meta OpenXR package.  I notice your Meta feature list is a little different than mine.  I'm using the com.unity.xr.meta-openxr package, version 1.0.1, which is what I thought we were supposed to use for Quest 3, for features like passthrough and such.  What package are you using?

image

But, it is interesting that the Quest 3 controller works in the legacy rig and I can switch between hand and controller no issue.  But in the new rig, with this change I think, the controller selection stops working for me (with the above config).  I'll go back and test if that is also the case before this change.

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs (hunk 5 above)
Comment (AMollis) [reply org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs]: Ahh yes :)  I made this better.   See the new helper function for details

**Comment by AMollis** [comment]
> @AMollis
> 
> Interesting, I enabled some of the features like you had, but no difference for me - still no controller selection, though it does track. Again, the controller totally works with the legacy rig, but for some reason not the new rig, for me. So you are seeing controller selection working in the new rig when switching from hands to controller?
> 
> image
> 
> I wonder if it has something to do with using a different meta OpenXR package. I notice your Meta feature list is a little different than mine. I'm using the com.unity.xr.meta-openxr package, version 1.0.1, which is what I thought we were supposed to use for Quest 3, for features like passthrough and such. What package are you using?
> 
> image
> 
> But, it is interesting that the Quest 3 controller works in the legacy rig and I can switch between hand and controller no issue. But in the new rig, with this change I think, the controller selection stops working for me (with the above config). I'll go back and test if that is also the case before this change.

@whebertML .... I was using the All-In-One package from Meta's publication on the Asset Store...per their documentation

https://assetstore.unity.com/packages/tools/integration/meta-xr-all-in-one-sdk-269657

**Comment by whebertML** [comment]
@AMollis That looks like a much more up to date package (June of 2024).  I searched some time ago and saw this package, https://docs.unity3d.com/Packages/com.unity.xr.meta-openxr@1.0/manual/index.html, and another document referencing it, took it for the latest info from Meta.  Guess not.  Thanks, will switch to using the latest one and see if it changes (I assume it will).

**Comment by AMollis** [comment]
> > @AMollis
> > Interesting, I enabled some of the features like you had, but no difference for me - still no controller selection, though it does track. Again, the controller totally works with the legacy rig, but for some reason not the new rig, for me. So you are seeing controller selection working in the new rig when switching from hands to controller?
> > image
> > I wonder if it has something to do with using a different meta OpenXR package. I notice your Meta feature list is a little different than mine. I'm using the com.unity.xr.meta-openxr package, version 1.0.1, which is what I thought we were supposed to use for Quest 3, for features like passthrough and such. What package are you using?
> > image
> > But, it is interesting that the Quest 3 controller works in the legacy rig and I can switch between hand and controller no issue. But in the new rig, with this change I think, the controller selection stops working for me (with the above config). I'll go back and test if that is also the case before this change.
> 
> @whebertML .... I was using the All-In-One package from Meta's publication on the Asset Store...per their documentation
> 
> https://assetstore.unity.com/packages/tools/integration/meta-xr-all-in-one-sdk-269657

I'll add in Untiy's [com.unity.xr.meta-openxr](https://docs.unity3d.com/Packages/com.unity.xr.meta-openxr@1.0/manual/index.html) and see what happens

**Comment by keveleigh** [comment]
> I was using the All-In-One package from Meta's publication on the Asset Store...per their documentation

For base support, we definitely _shouldn't_ need this (or even the Meta OpenXR package). In the past, all we needed was the `Meta Quest Support` feature and the `Oculus Touch Controller Profile` from Unity's base OpenXR Plugin package.

**Comment by whebertML** [comment]
Trying the all-in-one Meta package, I have other issues.  Like Hand selection not working in the legacy rig, though controller selection does.  But in the new rig, hands work, but the controller selection is "sticky", and hard to release.  Could be something odd with going back-and-forth in configuration.  Might need to start over clean - but may have to wait till next week for me.

**Comment by AMollis** [comment]
> t the controller selection is "sticky", an

Yes,... the sticky thing this is known.  I'll tackle that in a different PR.

**Comment by AMollis** [comment]
> > t the controller selection is "sticky", an
> 
> Yes,... the sticky thing this is known. I'll tackle that in a different PR.

This change at least get's quest 3 mostly working.   

We should get this change in and investigate other failures with additional PRs

**Comment by AMollis** [comment]
> > > t the controller selection is "sticky", an
> > 
> > 
> > Yes,... the sticky thing this is known. I'll tackle that in a different PR.
> 
> This change at least get's quest 3 mostly working with new rig. Minus the sticky behavior we're good.
> 
> We should get this change in and investigate other failures with additional PRs.
> 
> @whebertML is there something in particular you'd like to see changed with this PR? Happy to investigate other failures in a future PR. But this PR should be introducing any new regressions.

@whebertML  thanks for all the feedback :) 

I found the issue with the sticky select with Quest 3 controls.  It was a fatal flaw in my fix to the "PinchInputReader".  I can't just rely on selectInput.action.activeControl == null to determine if pitch should be driven by joint poses.   For actions action.activeControl is only non-null if the state was updated in the current frame. For buttons this only happens when you are actively pressing the button.  

As a result, the only stable way of determining if pinch should be using joint data is by querying the HandPoseDriver, and asking the HandPoseDriver if it is also using the joint poses directly (ie isPolyfilledPose).

This means I have to pass in the HandPoseDriver to the PinchInputReader...not horrible :)  And results in less code in the PinchInputReader.

**Review by shaynie — DISMISSED** [review]

**Comment by AMollis** [comment]
@whebertML ... please take a look at these latest changes.   I plan on merge this early tomorrow, to unblock the preview release.

**Review by shaynie — APPROVED** [review]

**Comment by AMollis** [comment]
> @whebertML ... please take a look at these latest changes. I plan on merging this early tomorrow, to unblock the preview release.

@whebertML  I'm going to merge this to unblock preview release.  I'll address additional feedback with future PR

**Comment by whebertML** [comment]
@AMollis Testing latest, but starting off with a clean slate with just latest feature/XRI3 (not yet pulling this PR) and no special Quest packages - just the base Meta Quest Support and Oculus Touch Controller Profile.  With that, I seem to have hands, hand rays, and Controllers working fine with both the legacy and new XRI3 rigs.

As @keveleigh suggests, seems things are fine at least with the base support.  So does that mean adding specific meta packages may, or just using specific profiles, has issues?  Will try a couple of combos.  Just wanting to be sure what issues the changes are actually intended to fix.

**Pull request merged**